AX: Voice control not focusing inputs and textareas when not in a fieldset

hoffmanjoshua opened the pull request:
#### 09204fce5fa15194dbec6ce3ad2d0772be628ab8
<pre>
AX: Voice control not focusing inputs and textareas when not in a fieldset
<a href="https://bugs.webkit.org/show_bug.cgi?id=293974">https://bugs.webkit.org/show_bug.cgi?id=293974</a>
<a href="https://rdar.apple.com/152517222">rdar://152517222</a>

Reviewed by Tyler Wilcock.

We expose text inputs as supporting press actions in supportsPressAction, but did
not handle them at all when computing a node&apos;s action element. This would lead ATs
like Voice Control to try to invoke this action, with no result or change in focus.

To start supporting this, return text inputs and content editable nodes as action
elements.

* LayoutTests/accessibility/mac/press-action-text-inputs-expected.txt: Added.
* LayoutTests/accessibility/mac/press-action-text-inputs.html: Added.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::nodeActionElement):

Canonical link: <a href="https://commits.webkit.org/296763@main">https://commits.webkit.org/296763@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e27b05fceaed31548ac21e8cf06e1049226cad9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109450 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29108 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19537 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114655 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59678 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111413 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29787 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37696 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83188 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112398 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23726 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98587 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63647 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23107 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16729 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59274 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93096 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16771 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117769 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36491 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27018 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92201 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36862 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94848 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92017 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23444 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36949 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14697 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32294 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36384 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41859 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36054 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39393 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37758 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1224,11 +1224,15 @@ static RefPtr<Element> nodeActionElement(Node& node)
         return &downcast<Element>(node);
 
+    // Content editable nodes should also be considered action elements, so they can accept presses.
+    if (HTMLElement::editabilityFromContentEditableAttr(node, PageIsEditable::No) != Node::Editability::ReadOnly)
```
This is a new ancestor walk. Is that necessary, or can we just check for the `contenteditable` attribute?

hoffmanjoshua replied:
Good catch—switched to using our existing our existing `contentEditableAttributeIsEnabled` static method.

minorninth commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1224,11 +1224,15 @@ static RefPtr<Element> nodeActionElement(Node& node)
     auto elementName = WebCore::elementName(node);
     if (RefPtr input = dynamicDowncast<HTMLInputElement>(node)) {
-        if (!input->isDisabledFormControl() && (input->isRadioButton() || input->isCheckbox() || input->isTextButton() || input->isFileUpload() || input->isImageButton()))
+        if (!input->isDisabledFormControl() && (input->isRadioButton() || input->isCheckbox() || input->isTextButton() || input->isFileUpload() || input->isImageButton() || input->isTextType()))
```
Can you double-check if isTextField might be more correct? The only difference is type=number, and I think that we may want to focus type=number on press too.

hoffmanjoshua replied:
Great call! Updated this, and added `type=number` to the test.

minorninth reviewed (approved):
lgtm with one possible change.

There are a lot of input types!

hoffmanjoshua committed:
AX: Voice control not focusing inputs and textareas when not in a fieldset
https://bugs.webkit.org/show_bug.cgi?id=293974
rdar://152517222

Reviewed by Tyler Wilcock.

We expose text inputs as supporting press actions in supportsPressAction, but did
not handle them at all when computing a node's action element. This would lead ATs
like Voice Control to try to invoke this action, with no result or change in focus.

To start supporting this, return text inputs and content editable nodes as action
elements.

* LayoutTests/accessibility/mac/press-action-text-inputs-expected.txt: Added.
* LayoutTests/accessibility/mac/press-action-text-inputs.html: Added.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::nodeActionElement):

Canonical link: https://commits.webkit.org/296763@main
